### ISSUE TITLE
It might be worth to use NIO with Java.

### ISSUE THREAD

**Issue by khennicb** [johnrjenson/JavaVsNode#3]
First, your article was really nice to read ! And I would love to read an other one like this one that gives some better chance to Java with the java.nio feature.

[java.nio (NIO stands for Non-blocking I/O) is a collection of Java programming language APIs that offer features for intensive I/O operations. It was introduced with the J2SE 1.4 release of Java by Sun Microsystems to complement an existing standard I/O. NIO was developed under the Java Community Process as JSR 51.](https://en.wikipedia.org/wiki/Non-blocking_I/O_(Java))

This feature was intruduced in July 7, 2011. I think it is stable now.

Do you think it's worth the try ?

Regards.

**Comment by johnrjenson** [comment]
@khennicb That's fair. I think I don't give enough credit to Java in my post for the non-blocking support that it provides. I think my opening statement that Java's IO is blocking is misleading. Even thought I do explain that it can be done at the end of the paragraph, I can see how the statement is contradictory. I'll update the wording in the post.

**Comment by johnrjenson** [comment]
If I have time in the future, it could be fun to experiment with something like Vert.x or Akka that process http requests in a non-blocking way, but they don't seem widely used in the industry, so that is why I simulated blocking IO for Java.

**Comment by khennicb** [comment]
It's nice to see that this project still matter for you !

Your opening statement is pretty clear about why you use blocking IO, and I understood why you made this choice. It's not your purpose to use low-level feature in java to achieve best performance.

But from the point of view of someone wondering what platform looks the best to start a project with good performance, it is an option to use thoses low-level features to boost IO performance. (especialy when IO is the main factor on performance).

I'm looking foward to see some new comparison now ! :-)

Regards,

**Comment by rafaduka** [comment]
The test would be between node vs vertx?

**Comment by johnrjenson** [comment]
Yeah I think that is what he is basically asking, but the comparison wouldn't really tell us anything that we didn't already know. I think it is clear that Java would beat Node in that case.

The larger and more difficult problem to solve is that the Java ecosystem is really a blocking one. If Java fully embraced non-blocking io in every way and had a vast array of well supported NIO replacements for all of the widely used libraries in the community, then I think Node quite possibly may not have ever become a thing in the first place.

Read the section called "A Non-Blocking Ecosystem Matters".
https://www.tandemseven.com/blog/node-a-compelling-choice-for-java-developers/

**Comment by StevenCurran** [comment]
It already does. Netty. Would be interesting to see if you made a more fair comparison. Of course something in a busy event loop will beat creating 100 threads in java. Just because java allows it, and has for many years, doesn't mean you should. It's an apples to oranges comparison. 

I cam across this post to fairly compare writing a backend in node, vs what I'm used to in java. But the comparison is void considering the approach on the java side.

**Comment by varjasz** [comment]
Not a fair comparison.

Replace that java code:
```
		try(RandomAccessFile aFile = new RandomAccessFile(pathToFile, "r"); FileChannel inChannel = aFile.getChannel()) {
			long fileSize = inChannel.size();
			ByteBuffer buffer = ByteBuffer.allocate((int) fileSize);
			inChannel.read(buffer);
			buffer.flip();
			StringBuilder sb = new StringBuilder();
			for (int i = 0; i < fileSize; i++) {
				sb.append((char) buffer.get());
			}
			return sb.toString();
		} catch (IOException e) {
			throw e;
		}

```

with this:
`	return Files.readString(Path.of(pathToFile));
`

The file operations will be 6! times faster.

**Comment by brunoborges** [comment]
@johnrjenson are you still maintaining this?

**Comment by johnrjenson** [comment]
No I’m not.

John Jenson
m 857.202.1180    w rightpoint.com

RIGHTPOINT
a genpact company

**Comment by ashu2012** [comment]
any idea whats the idle spring boot java resquest per second metric ? i read it should be 10000 but with nio i read whatsapp is ale to achieve 1lk request per second.
